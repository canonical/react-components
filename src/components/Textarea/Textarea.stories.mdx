import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import Textarea from "./Textarea";

<Meta
  title="Textarea"
  component={Textarea}
  argTypes={{
    caution: {
      control: {
        type: "text",
      },
    },
    error: {
      control: {
        type: "text",
      },
    },
    help: {
      control: {
        type: "text",
      },
    },
    label: {
      control: {
        type: "text",
      },
    },
    success: {
      control: {
        type: "text",
      },
    },
    id: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => <Textarea {...args} />;

### Textarea

This is a [React](https://reactjs.org/) component for the Vanilla [Textarea](https://docs.vanillaframework.io/base/forms/#textarea).

The Textarea component defines a multi-line text input control.

### Props

<ArgsTable of={Textarea} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      label: "Tell us about your project or interest",
      id: "textarea2",
      rows: "3",
      defaultValue: "Ubuntu",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Dynamic Height

<Canvas>
  <Story
    name="Dynamic Height"
    args={{
      label: "Text area that will dynamically adjust its height based on content",
      defaultValue: "Ubuntu",
      grow: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
