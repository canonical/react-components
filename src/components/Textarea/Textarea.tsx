import classNames from "classnames";
import React, {
  useCallback,
  useEffect,
  useId,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import type { TextareaHTMLAttributes, ReactNode } from "react";

import Field from "../Field";

import type { ClassName, PropsWithSpread } from "types";

/**
 * The props for the Textarea component.
 */
export type Props = PropsWithSpread<
  {
    /**
     * The content for caution validation.
     */
    caution?: ReactNode;
    /**
     * Optional class(es) to pass to the textarea element.
     */
    className?: ClassName;
    /**
     * The content for error validation.
     */
    error?: ReactNode;
    /**
     * Whether the textarea should grow to fit the content automatically.
     */
    grow?: boolean;
    /**
     * Help text to show below the field.
     */
    help?: ReactNode;
    /**
     * The id of the textarea.
     */
    id?: string;
    /**
     * The label for the field.
     */
    label?: ReactNode;
    /**
     * Optional class(es) to pass to the label component.
     */
    labelClassName?: string;
    /**
     * Whether the field is required.
     */
    required?: boolean;
    /**
     * Whether the form field should have a stacked appearance.
     */
    stacked?: boolean;
    /**
     * The content for success validation.
     */
    success?: ReactNode;
    /**
     * Whether to focus on the input on initial render.
     */
    takeFocus?: boolean;
    /**
     * Optional class(es) to pass to the wrapping Field component
     */
    wrapperClassName?: string;
    /**
     * Function to occur on keyboard event, 'CTRL + Enter'
     */
    onControlEnter?: () => void;
  },
  TextareaHTMLAttributes<HTMLTextAreaElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Textarea](https://docs.vanillaframework.io/base/forms/#textarea).
 *
 * The Textarea component defines a multi-line text input control.
 */
const Textarea = ({
  caution,
  className,
  error,
  grow = false,
  help,
  id,
  label,
  labelClassName,
  onKeyUp,
  onControlEnter,
  required,
  stacked,
  style,
  success,
  takeFocus = false,
  wrapperClassName,
  ...props
}: Props): JSX.Element => {
  const textareaRef = useRef(null);
  const validationId = useId();
  const helpId = useId();
  const hasError = !!error;
  const [innerValue, setInnervalue] = useState(props.defaultValue);
  const defaultTextAreaId = useId();
  const textAreaId = id || defaultTextAreaId;

  const handleKeyDown = useCallback(
    (event: KeyboardEvent) => {
      if (
        event.key === "Enter" &&
        (event.ctrlKey || event.metaKey) &&
        document.activeElement === textareaRef.current
      ) {
        onControlEnter();
      }
    },
    [onControlEnter]
  );

  useEffect(() => {
    document.addEventListener("keydown", handleKeyDown);

    return () => {
      document.removeEventListener("keydown", handleKeyDown);
    };
  }, [handleKeyDown]);

  useEffect(() => {
    if (takeFocus) {
      textareaRef.current.focus();
    }
  }, [takeFocus]);

  useLayoutEffect(() => {
    if (grow) {
      const textArea = textareaRef.current;
      if (textArea) {
        textArea.style.height = "0px";
        const scrollHeight = textArea.scrollHeight;
        textArea.style.height = `${scrollHeight}px`;
      }
    }
  }, [textareaRef, grow, innerValue, props.value]);

  return (
    <Field
      caution={caution}
      className={wrapperClassName}
      error={error}
      forId={textAreaId}
      help={help}
      helpId={helpId}
      label={label}
      labelClassName={labelClassName}
      required={required}
      stacked={stacked}
      success={success}
      validationId={validationId}
    >
      <textarea
        aria-describedby={[help ? helpId : null, success ? validationId : null]
          .filter(Boolean)
          .join(" ")}
        aria-errormessage={hasError ? validationId : null}
        aria-invalid={hasError}
        className={classNames("p-form-validation__input", className)}
        id={textAreaId}
        onKeyUp={(evt) => {
          onKeyUp && onKeyUp(evt);
        }}
        onChange={(evt) => {
          if (!props.value) {
            setInnervalue(evt.target.value);
          }
        }}
        ref={textareaRef}
        style={
          (grow && {
            minHeight: "5rem",
            resize: "none",
            overflow: "hidden",
            boxSizing: "border-box",
            ...style,
          }) ||
          style
        }
        {...props}
        value={props.value ?? innerValue}
      />
    </Field>
  );
};

export default Textarea;
