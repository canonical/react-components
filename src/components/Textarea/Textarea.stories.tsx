import { Meta, StoryObj } from "@storybook/react";

import Textarea from "./Textarea";

const meta: Meta<typeof Textarea> = {
  component: Textarea,
  tags: ["autodocs"],

  argTypes: {
    caution: {
      control: {
        type: "text",
      },
    },

    error: {
      control: {
        type: "text",
      },
    },

    help: {
      control: {
        type: "text",
      },
    },

    label: {
      control: {
        type: "text",
      },
    },

    success: {
      control: {
        type: "text",
      },
    },

    id: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Textarea>;

export const Default: Story = {
  name: "Default",

  args: {
    label: "Tell us about your project or interest",
    id: "textarea2",
    rows: 3,
    defaultValue: "Ubuntu",
  },
};

export const DynamicHeight: Story = {
  name: "Dynamic Height",

  args: {
    label: "Text area that will dynamically adjust its height based on content",
    defaultValue: "Ubuntu",
    grow: true,
  },
};
