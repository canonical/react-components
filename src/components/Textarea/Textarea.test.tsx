import { mount, shallow } from "enzyme";
import React from "react";

import Textarea from "./Textarea";

describe("Textarea ", () => {
  it("renders", () => {
    const wrapper = shallow(<Textarea id="test-id" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("can add additional classes", () => {
    const wrapper = shallow(<Textarea className="extra-class" />);
    const className = wrapper.find("textarea").prop("className");
    expect(className.includes("p-form-validation__input")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });

  it("can take focus on first render", () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    const wrapper = mount(<Textarea takeFocus />, { attachTo: container });
    expect(wrapper.find("textarea").getDOMNode()).toBe(document.activeElement);
  });

  it("can accept input attributes", () => {
    const wrapper = shallow(<Textarea autoComplete="on" />);
    expect(wrapper.exists()).toBe(true);
  });
});
