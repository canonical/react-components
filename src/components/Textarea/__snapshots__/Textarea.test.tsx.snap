// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea  renders 1`] = `
<div
  class="p-form__group p-form-validation textarea"
>
  <div
    class="p-form__control u-clearfix"
  >
    <textarea
      aria-describedby=""
      aria-invalid="false"
      class="p-form-validation__input"
      id="test-id"
    />
  </div>
</div>
`;
